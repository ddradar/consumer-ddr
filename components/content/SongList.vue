<template>
  <OTable :data="songs!" striped narrowed :loading="isLoading">
    <OTableColumn v-slot="props" field="name" label="Name">
      <NuxtLink :to="`/songs/${props.row.slug}/`">
        {{ props.row.name }}
      </NuxtLink>
    </OTableColumn>
    <OTableColumn v-slot="props" field="artist" label="Artist">
      <span>{{ props.row.artist }}</span>
    </OTableColumn>
    <OTableColumn v-slot="props" field="bpm" label="BPM">
      {{ props.row.bpm }}
    </OTableColumn>
    <OTableColumn
      v-for="playStyle in playStyles"
      v-slot="props"
      :key="playStyle"
      :label="playStyle"
    >
      <div class="tags has-addons">
        <span
          v-for="chart in props.row.charts[playStyle]"
          :key="chart.difficulty"
          class="tag"
          :class="chart.color"
          :title="chart.name"
        >
          {{ chart.level }}
        </span>
      </div>
    </OTableColumn>

    <template #empty>
      <section class="section">
        <div class="content has-text-grey has-text-centered">
          <p>Nothing here.</p>
        </div>
      </section>
    </template>
  </OTable>
</template>

<script lang="ts" setup>
import type {
  Chart,
  PlayStyle,
  SoftwareParsedContent,
  SongParsedContent
} from '~~/src/content'
import { normalizeDifficulty } from '~~/src/song'

type ChartInfo = Pick<Chart, 'level' | 'difficulty'> & {
  color: string
  name: string
}
interface Props {
  series: string
}

const prop = defineProps<Props>()

const { data: software } = await useAsyncData(`/software/${prop.series}`, () =>
  queryContent<SoftwareParsedContent>(prop.series)
    .where({ _type: 'markdown' })
    .findOne()
)

const { data: songs, pending: isLoading } = await useAsyncData(
  `/software/${prop.series}/songs`,
  () =>
    queryContent<SongParsedContent>(prop.series)
      .where({ _type: 'json' })
      .without('series')
      .find(),
  {
    transform: (songs: SongParsedContent[]) =>
      songs.map((s) => ({
        slug: s.slug,
        name: s.name,
        artist: s.artist,
        bpm: s.bpm,
        charts: s.charts.reduce((p, c) => {
          if (!p[c.playStyle]) {
            p[c.playStyle] = []
          }
          const name =
            software.value?.difficultyNames[c.difficulty] ?? 'Unknown'
          p[c.playStyle]!.push({
            level: c.level,
            difficulty: c.difficulty,
            name,
            color: `is-${normalizeDifficulty(name).toLowerCase()}`
          })
          return p
        }, {} as Partial<Record<PlayStyle, ChartInfo[]>>)
      }))
  }
)

const playStyles = computed(() => [
  ...new Set(songs.value?.flatMap((s) => Object.keys(s.charts)))
])
</script>
