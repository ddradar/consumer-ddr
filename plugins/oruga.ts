import Oruga from '@oruga-ui/oruga-next'
import { bulmaConfig } from '@oruga-ui/theme-bulma'

export default defineNuxtPlugin(({ vueApp }) => {
  vueApp.use(Oruga, bulmaConfig)
})

declare module '@vue/runtime-core' {
  export interface GlobalComponents {
    OAutocomplete: typeof import('@oruga-ui/oruga-next/src/components/autocomplete')['OAutocomplete']
    OLoading: typeof import('@oruga-ui/oruga-next/src/components/loading')['OLoading']
    OTable: typeof import('@oruga-ui/oruga-next/src/components/table')['OTable']
    OTableColumn: typeof import('@oruga-ui/oruga-next/src/components/table')['OTableColumn']
    OTooltip: typeof import('@oruga-ui/oruga-next/src/components/tooltip')['OTooltip']
  }
}
