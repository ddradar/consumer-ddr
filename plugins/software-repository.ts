import softwareList from '~/series'
import { Chart } from '~/types/chart'
import { Software } from '~/types/software'
import { Song } from '~/types/song'

export const getSoftwareList = (): Software[] =>
  softwareList.map(
    (s) =>
      ({
        id: s.id,
        name: s.name,
        region: s.region,
        platform: s.platform,
        launched: new Date(s.launched)
      } as Software)
  )

type SongWithCharts = Song & { charts: Chart[] }
type SoftwareInfo = Software & { songs: SongWithCharts[] }
export const getSoftwareInfo = (id: string): SoftwareInfo | undefined => {
  const info = softwareList.find((s) => s.id === id)
  if (info === undefined) return info
  return {
    id: info.id,
    name: info.name,
    region: info.region,
    platform: info.platform,
    launched: new Date(info.launched),
    songs: info.songs.map((s) => ({
      id: s.id,
      name: s.name,
      artist: s.artist,
      bpm: s.bpm,
      charts: s.charts.map((c) => ({
        songId: s.id,
        playStyle: c.playStyle,
        difficulty: c.difficulty,
        level: c.level
      }))
    }))
  } as SoftwareInfo
}
