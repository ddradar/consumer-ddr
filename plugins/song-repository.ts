import softwareList from '../series'
import { Chart } from '../types/chart'
import { Song } from '../types/song'

type ChartItem = Omit<Chart, 'level'> & { level: (number | '10+')[] }
type SongInfo = Song & {
  charts: ChartItem[]
  softwareIds: string[]
}

export const getSongList = (): SongInfo[] =>
  softwareList.reduce((prev, current) => {
    for (const song of current.songs) {
      const songInfo = prev.find((s) => s.id === song.id)
      if (songInfo) {
        songInfo.softwareIds.push(current.id)
        for (const chart of song.charts) {
          const chartInfo = songInfo.charts.find(
            (c) =>
              c.playStyle === chart.playStyle &&
              c.difficulty === chart.difficulty
          )
          if (chartInfo && !chartInfo.level.find((l) => l === chart.level)) {
            chartInfo.level.push(chart.level)
          } else {
            songInfo.charts.push({
              ...chart,
              songId: song.id,
              level: [chart.level]
            })
          }
        }
      } else {
        prev.push({
          ...song,
          charts: song.charts.map((c) => ({
            ...c,
            songId: song.id,
            level: [c.level]
          })),
          softwareIds: [current.id]
        })
      }
    }
    return prev
  }, [] as SongInfo[])
