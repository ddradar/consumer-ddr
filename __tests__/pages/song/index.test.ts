import {
  createLocalVue,
  mount,
  RouterLinkStub,
  shallowMount,
  Wrapper
} from '@vue/test-utils'
import Buefy from 'buefy'
import { mocked } from 'ts-jest/utils'
import { MetaInfo } from 'vue-meta'

import SongList from '~/pages/song/index.vue'
import { getSoftwareList } from '~/plugins/software-repository'
import { getSongList } from '~/plugins/song-repository'

const localVue = createLocalVue()
localVue.use(Buefy)
jest.mock('~/plugins/software-repository')
jest.mock('~/plugins/song-repository')
mocked(getSoftwareList).mockReturnValue([
  {
    id: '1st-jp',
    name: 'Dance Dance Revolution',
    region: 'JP',
    platform: 'Play Station',
    launched: new Date('1999-04-10')
  },
  {
    id: '2nd-remix-jp',
    name: 'Dance Dance Revolution 2nd ReMIX',
    platform: 'Play Station',
    region: 'JP',
    launched: new Date('1999-08-26')
  }
])
mocked(getSongList).mockReturnValue([
  {
    id: 'am-3p',
    name: 'AM-3P',
    artist: 'KTz',
    bpm: 130,
    softwareIds: ['2nd-remix-jp'],
    charts: [
      {
        playStyle: 'SINGLE',
        difficulty: 1,
        level: [5]
      },
      {
        playStyle: 'SINGLE',
        difficulty: 2,
        level: [6]
      },
      {
        playStyle: 'SINGLE',
        difficulty: 3,
        level: [8]
      },
      {
        playStyle: 'DOUBLE',
        difficulty: 1,
        level: [5]
      },
      {
        playStyle: 'DOUBLE',
        difficulty: 2,
        level: [6]
      }
    ]
  },
  {
    id: 'paranoia',
    name: 'PARANOiA',
    artist: '180',
    bpm: 180,
    softwareIds: ['1st-jp', '2nd-remix-jp'],
    charts: [
      {
        playStyle: 'SINGLE',
        difficulty: 1,
        level: [6]
      },
      {
        playStyle: 'SINGLE',
        difficulty: 2,
        level: [7]
      }
    ]
  },
  {
    id: 'paranoia',
    name: 'PARANOiA',
    artist: '180',
    bpm: 180,
    softwareIds: ['1st-jp', '2nd-remix-jp'],
    charts: [
      {
        playStyle: 'SINGLE',
        difficulty: 1,
        level: [6]
      },
      {
        playStyle: 'SINGLE',
        difficulty: 2,
        level: [7]
      }
    ]
  }
])

describe('pages/song/index.vue', () => {
  let wrapper: Wrapper<Vue>

  beforeEach(() => {
    wrapper = shallowMount(SongList, {
      localVue,
      stubs: {
        NuxtLink: RouterLinkStub
      }
    })
  })

  test('renders correctly', () => {
    const wrapper = mount(SongList, {
      localVue,
      stubs: {
        NuxtLink: RouterLinkStub
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
  describe('head', () => {
    test('returns "Song List"', () => {
      // Arrange
      const head = wrapper.vm.$data.head as MetaInfo

      // Act & Assert
      expect(head).toStrictEqual({ title: 'Song List' })
    })
  })
})
