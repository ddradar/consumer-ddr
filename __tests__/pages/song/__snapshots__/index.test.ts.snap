// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/song/index.vue renders correctly 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    Song List
  </h1>
   
  <div
    class="b-table"
  >
    <!---->
     
    <!---->
     
    <!---->
     
    <!---->
    <!---->
    <!---->
    <div
      class="table-wrapper has-mobile-cards"
    >
      <table
        class="table is-striped is-narrow"
      >
        <thead>
          <tr>
            <!---->
            <!---->
            <th
              class=""
              draggable="false"
            >
              <div
                class="th-wrap"
              >
                <span
                  class="is-relative"
                >
                   Name 
                  <span
                    class="icon sort-icon is-small is-invisible"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </div>
            </th>
            <th
              class=""
              draggable="false"
            >
              <div
                class="th-wrap"
              >
                <span
                  class="is-relative"
                >
                   Artist 
                  <span
                    class="icon sort-icon is-small is-invisible"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </div>
            </th>
            <th
              class=""
              draggable="false"
            >
              <div
                class="th-wrap"
              >
                <span
                  class="is-relative"
                >
                   BPM 
                  <span
                    class="icon sort-icon is-small is-invisible"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </div>
            </th>
            <th
              class=""
              draggable="false"
            >
              <div
                class="th-wrap"
              >
                <span
                  class="is-relative"
                >
                   Series 
                  <span
                    class="icon sort-icon is-small is-invisible"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </div>
            </th>
            <!---->
          </tr>
          <!---->
          <tr>
            <!---->
            <!---->
            <th
              class=""
            >
              <div
                class="th-wrap"
              >
                <div
                  class="control is-clearfix"
                >
                  <input
                    autocomplete="on"
                    class="input"
                    type="text"
                  />
                  <!---->
                  <!---->
                  <!---->
                </div>
              </div>
            </th>
            <th
              class=""
            >
              <div
                class="th-wrap"
              >
                <div
                  class="control is-clearfix"
                >
                  <input
                    autocomplete="on"
                    class="input"
                    type="text"
                  />
                  <!---->
                  <!---->
                  <!---->
                </div>
              </div>
            </th>
            <th
              class=""
            >
              <div
                class="th-wrap"
              >
                <!---->
              </div>
            </th>
            <th
              class=""
            >
              <div
                class="th-wrap"
              >
                <!---->
              </div>
            </th>
            <!---->
          </tr>
        </thead>
        <tbody>
          <tr
            class=""
            draggable="false"
          >
            <!---->
            <!---->
            <td
              class=""
              data-label="Name"
            >
              <a
                class="is-size-6-mobile"
              >
                
        AM-3P
      
              </a>
            </td>
            <td
              class=""
              data-label="Artist"
            >
              <span
                class="is-size-6-mobile"
              >
                KTz
              </span>
            </td>
            <td
              class=""
              data-label="BPM"
            >
              
      130
    
            </td>
            <td
              class=""
              data-label="Series"
            >
              <div
                class="buttons"
              >
                <a
                  class="button is-small is-success"
                  type="button"
                >
                  <!---->
                  <span>
                    
          2nd-remix-jp
        
                  </span>
                  <!---->
                </a>
              </div>
            </td>
            <!---->
          </tr>
          <transition-stub
            name=""
          >
            <!---->
          </transition-stub>
          <!---->
          <tr
            class=""
            draggable="false"
          >
            <!---->
            <!---->
            <td
              class=""
              data-label="Name"
            >
              <a
                class="is-size-6-mobile"
              >
                
        PARANOiA
      
              </a>
            </td>
            <td
              class=""
              data-label="Artist"
            >
              <span
                class="is-size-6-mobile"
              >
                180
              </span>
            </td>
            <td
              class=""
              data-label="BPM"
            >
              
      180
    
            </td>
            <td
              class=""
              data-label="Series"
            >
              <div
                class="buttons"
              >
                <a
                  class="button is-small is-info"
                  type="button"
                >
                  <!---->
                  <span>
                    
          1st-jp
        
                  </span>
                  <!---->
                </a>
                <a
                  class="button is-small is-success"
                  type="button"
                >
                  <!---->
                  <span>
                    
          2nd-remix-jp
        
                  </span>
                  <!---->
                </a>
              </div>
            </td>
            <!---->
          </tr>
          <transition-stub
            name=""
          >
            <!---->
          </transition-stub>
          <!---->
          <!---->
        </tbody>
        <!---->
      </table>
      <!---->
    </div>
    <div
      class="top level"
    >
      <div
        class="level-left"
      />
      <div
        class="level-right"
      >
        <div
          class="level-item"
        >
          <nav
            class="pagination"
          >
            <a
              class="pagination-link pagination-previous"
              disabled="disabled"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-left mdi-24px"
                />
              </span>
            </a>
            <a
              class="pagination-link pagination-next"
              disabled="disabled"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-right mdi-24px"
                />
              </span>
            </a>
            <ul
              class="pagination-list"
            >
              <!---->
              <!---->
              <li>
                <a
                  aria-current="true"
                  class="pagination-link is-current "
                  href="#"
                  role="button"
                >
                  1
                </a>
              </li>
              <!---->
              <!---->
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </div>
</section>
`;
